### ISSUE TITLE
Error using ActionBarDrawerToggle in appcompat-v7

### ISSUE THREAD

**Issue by ajhuntsman** [robolectric/robolectric#1644]
I'm using 3.0-rc1 so I'm wondering if this is a duplicate of #1424 or possibly a regression from #1446 somehow?

Every unit test fails with this same exception:

```
java.lang.RuntimeException: expected android:id/home to be a attr, is a id
	at org.robolectric.res.ResName.mustBe(ResName.java:130)
	at org.robolectric.shadows.ShadowAssetManager$StyleResolver.getAttrValue(ShadowAssetManager.java:408)
	at org.robolectric.shadows.ShadowResources.getOverlayedThemeValue(ShadowResources.java:291)
	at org.robolectric.shadows.ShadowResources.findAttributeValue(ShadowResources.java:284)
	at org.robolectric.shadows.ShadowResources.attrsToTypedArray(ShadowResources.java:187)
	at org.robolectric.shadows.ShadowResources.access$000(ShadowResources.java:51)
	at org.robolectric.shadows.ShadowResources$ShadowTheme.obtainStyledAttributes(ShadowResources.java:489)
	at org.robolectric.shadows.ShadowResources$ShadowTheme.obtainStyledAttributes(ShadowResources.java:484)
	at org.robolectric.shadows.ShadowResources$ShadowTheme.obtainStyledAttributes(ShadowResources.java:479)
	at android.content.res.Resources$Theme.obtainStyledAttributes(Resources.java)
	at android.content.Context.obtainStyledAttributes(Context.java:436)
	at android.support.v7.app.ActionBarDrawerToggle$ToolbarCompatDelegate.getThemeUpIndicator(ActionBarDrawerToggle.java:621)
	at android.support.v7.app.ActionBarDrawerToggle.getThemeUpIndicator(ActionBarDrawerToggle.java:460)
	at android.support.v7.app.ActionBarDrawerToggle.__constructor__(ActionBarDrawerToggle.java:228)
	at android.support.v7.app.ActionBarDrawerToggle.<init>(ActionBarDrawerToggle.java)
	at org.acme.acmeapp.activity.MainActivity.onCreate(MainActivity.java:586)
	at android.app.Activity.performCreate(Activity.java:5933)
```

**Comment by erd** [comment]
I'll try and take a look at this.

**Comment by nenick** [comment]
@ajhuntsman you can use the workaround i have posted at https://github.com/robolectric/robolectric/issues/1446 to avoid this error

**Comment by ajhuntsman** [comment]
@nenick - I tried it, but that didn't work. Does it matter which package the custom class is in? Do I have to register it somewhere, or does Robolectric just pick it up via annotation scanning?

**Comment by mehmetf** [comment]
I just ran into this issue as well. I will try @nenick 's workaround. Shadows need to be registered by referencing them in your config. Something like:

@Config(manifest = "/somepath/AndroidManifest.xml",
    shadows = {MyShadow.class})

It seems the issue is related to toolbar. As evident from your stack trace (and mine) this is happening with ToolbarCompatDelegate.

**Comment by mehmetf** [comment]
@nenick 's workaround solves this one as well. Verified, thanks!

**Comment by ajhuntsman** [comment]
Thanks @mehmetf for the clarification. The workaround is working for me, but here are some additional comments for posterity.

I'm using the `robolectric.properties` file already, to declare my manifest location and API levels. I was able to register my shadow using the **fully qualified** class name, but **without** the ".class" suffix:
```
emulateSdk=21
reportSdk=21
manifest=./src/main/AndroidManifest.xml
constants=org.acme.acmeapp.BuildConfig
shadows=org.acme.acmeapp.test.ShadowActionBarDrawerToggle
```
Alternatively, the same thing can be accomplished by registering the shadow via the `@Config` annotation, but this requires the **unqualified** class name **with** the ".class" suffix:
```
@Config(constants = BuildConfig.class, shadows = {ShadowActionBarDrawerToggle.class})
```
I'm going to leverage the `robolectric.properties` file to avoid annotation sprawl among my test classes. This means I no longer need the `@Config` annotation at all, for any test class. I really like having all of my Robolectric parameters in a single file.

Now if the `@RunWith` annotation could be defined in `robolectric.properties` that would be great!

**Comment by erd** [comment]
@ajhuntsman If you are using the Gradle runner, you do not need to specify the `manifest` path. The Gradle runner ignores it, actually. Also, I would suggest not specifying `emulateSdk` or `reportSdk` since those values will be read from your manifest.

**Comment by ajhuntsman** [comment]
Thanks for the clarification, @erd. Since I switched over to the Gradle runner, I didn't realize I should cleanup my configuration.

My new `robolectric.properties` file looks like this, and it's working just fine:
```
constants=org.acme.acmeapp.BuildConfig
shadows=org.acme.acmeapp.test.ShadowActionBarDrawerToggle
```

**Comment by erd** [comment]
Fantastic!

**Comment by erd** [comment]
I updated the `android-appcompat-v7` sample in `robolectric-samples` to use a DrawerLayout and ActionBarDrawerToggle and it works without any workarounds. @ajhuntsman - Could you take a look at it and see what's different vs your environment?

**Comment by erd** [comment]
I'm closing this as I'm unable to reproduce it. If you're able to reproduce this using the `robolectric-samples/android-appcompat-v7` project, please let me know.

**Comment by mehmetf** [comment]
This only happens if you use the particular constructor of actionbardrawertoggle that allows you to pass a toolbar in. If you use the common ctr, you will not see a problem.

**Comment by mehmetf** [comment]
This only happens if you use the particular constructor of
actionbardrawertoggle that allows you to pass a toolbar in. If you use the
common ctr, you will not see a problem.

On Friday, April 17, 2015, Erich Douglass <notifications@github.com> wrote:

> Closed #1644 <https://github.com/robolectric/robolectric/issues/1644>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/robolectric/robolectric/issues/1644#event-284047624>.
>


-- 
- Mehmet

**Comment by erd** [comment]
The sample mentioned above does use the `Toolbar` constructor.
